raster-foundry/raster-foundry
# Add caching logic to projects, project layers, and users

## Overview

This issue is in pursuit of better tile server performance. A common query that happens in the tile server is requesting a project and project layer by ID. This PR adds support to the DAOs for caching selects by ID and deleting the key when an update or delete is called in the DAO.

Along the way I added support for `User`s as well since that is a common query that happens all over the place in the tile server (and API).

### Checklist


### Notes

Caching `IO[Option[_]]` is kind of a pain - out of the box it will never get cached. @jisantuc had an idea that there was a way to define `Mode` for that, but I couldn't figure it out. Instead, I defined a [function](https://github.com/raster-foundry/raster-foundry/pull/5144/files#diff-10b547bf0417282ce29d0c4d44f0b020R21) that handles this caching and is reused -- it's not too bad, but if there's a better way to do it I'm open to suggestions.

## Testing Instructions

This tests a few aspects of caching _generally_ by specifically exercising common functionality across the caching in `UserDao`, `ProjectDao`, and `ProjectLayerDao` in a limited way rather than an expansive way since that would have been really hard to verify manually.
 - Reassemble and start your backsplash servers, also restart memcached to make sure to start from scratch for the cache `docker-compose restart memcached`
 - Make a `PUT` request - this was an issue during development because `memcached` returns an error, here we want to make sure with an empty cache an update that tries to delete a key doesn't fail
```
  export JWT_AUTH_TOKEN=<auth-token>

  echo '{"id":"410d4975-0eed-40c5-aa5a-511277507191","createdAt":"2019-01-21T17:56:41.247Z","modifiedAt":"2019-08-28T16:44:43.564Z","createdBy":"auth0|59318a9d2fbbca3e16bcfc92","owner":"auth0|59318a9d2fbbca3e16bcfc92","name":"Local File Single Scene","slugLabel":"Local File Single Scene","description":"","visibility":"PUBLIC","tileVisibility":"PUBLIC","isAOIProject":false,"aoiCadenceMillis":604800000,"aoisLastChecked":"2019-01-21T17:56:41.247Z","tags":["test"],"extent":{"type":"Polygon","coordinates":[[[-100.81800544812673,41.12110187268947],[-100.81800544812673,41.191357235936],[-100.7445369637344,41.191357235936],[-100.7445369637344,41.12110187268947],[-100.81800544812673,41.12110187268947]]]},"manualOrder":true,"isSingleBand":false,"singleBandOptions":null,"defaultAnnotationGroup":null,"extras":null,"defaultLayerId":"301991bb-6198-48d3-ad87-8fc8bf5a67d9"}' | http --auth-type=jwt PUT :9100/api/projects/410d4975-0eed-40c5-aa5a-511277507191
```
  - Make two `GET`requests and examine the SQL logs for the API server - there should be one query on the first, none on the second for projects
```
  http --auth-type=jwt :9100/api/projects/410d4975-0eed-40c5-aa5a-511277507191
```
 - Make multiple tile requests for a tile from the project, on the first request no SQL query for the project should happen, but one should happen for the project layer and user. On the second request no queries for the project, project layer, or user should happen
```
  http ":8081/410d4975-0eed-40c5-aa5a-511277507191/16/14422/24531/?tag=1567513934821&token=$JWT_AUTH_TOKEN" > /dev/null
```

Closes #5120
